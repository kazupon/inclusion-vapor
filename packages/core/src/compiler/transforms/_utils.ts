import { parse } from 'svelte/compiler'
import { generate } from '@vue-vapor/compiler-vapor'
import { transform } from '../transform'
import { IRNodeTypes } from '../ir'

import type {
  RootIRNode as VaporRootIRNode,
  CompilerOptions as VaporCompilerOptions
} from '@vue-vapor/compiler-vapor'
import type { CompilerOptions } from '../compile'
import type { RootNode } from '../ir'

export const DEFAULT_OPTIONS: CompilerOptions = {
  prefixIdentifiers: true
}
export const DEFAULT_VAPOR_COMPILER_OPTIONS = DEFAULT_OPTIONS as VaporCompilerOptions

export function makeCompile(options: CompilerOptions = {}) {
  return (source: string, overrideOptions: CompilerOptions = {}) => {
    const svelteAst = parse(source)

    const ast: RootNode = {
      type: IRNodeTypes.ROOT,
      children: svelteAst.html.children || [],
      source,
      components: [],
      directives: [],
      helpers: new Set(),
      temps: 0
    }

    const ir = transform(ast, {
      ...DEFAULT_OPTIONS,
      ...options,
      ...overrideOptions
    })
    const { code, helpers, vaporHelpers } = generate(ir as unknown as VaporRootIRNode, {
      ...DEFAULT_OPTIONS,
      ...options,
      ...overrideOptions
    })

    return { ast, ir, code, helpers, vaporHelpers }
  }
}
