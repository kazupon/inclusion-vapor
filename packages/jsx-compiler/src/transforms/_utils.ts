import { parse } from '@babel/parser'
import { generate } from '@vue-vapor/compiler-vapor'
import { transform } from '../transform'
import { IRNodeTypes } from '../ir'

import type {
  RootIRNode as VaporRootIRNode,
  CompilerOptions as VaporCompilerOptions
} from '@vue-vapor/compiler-vapor'
import type { CompilerOptions } from '../compile'
import type { RootNode, JSXElement, JSXFragment } from '../ir'

export const DEFAULT_OPTIONS: CompilerOptions = {
  prefixIdentifiers: true
}
export const DEFAULT_VAPOR_COMPILER_OPTIONS = DEFAULT_OPTIONS as VaporCompilerOptions

export function makeCompile(options: CompilerOptions = {}) {
  return (source: string, overrideOptions: CompilerOptions = {}) => {
    const {
      body: [statement]
    } = parse(source, {
      sourceType: 'module',
      plugins: ['jsx']
    }).program
    let children!: JSXElement[] | JSXFragment['children']
    if (statement.type === 'ExpressionStatement') {
      children =
        statement.expression.type === 'JSXFragment'
          ? statement.expression.children
          : (statement.expression.type === 'JSXElement'
            ? [statement.expression]
            : [])
    }

    const ast: RootNode = {
      type: IRNodeTypes.ROOT,
      children,
      source,
      components: [],
      directives: [],
      helpers: new Set(),
      temps: 0
    }

    const ir = transform(ast, {
      ...DEFAULT_OPTIONS,
      ...options,
      ...overrideOptions
    })
    const { code, helpers, vaporHelpers } = generate(ir as unknown as VaporRootIRNode, {
      ...DEFAULT_OPTIONS,
      ...options,
      ...overrideOptions
    })

    return { ast, ir, code, helpers, vaporHelpers }
  }
}
