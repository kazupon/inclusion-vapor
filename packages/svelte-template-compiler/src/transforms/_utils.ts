import { parse } from 'svelte/compiler'
import { generate } from '@vue-vapor/compiler-vapor'
import { transform } from '../transform'
import { IRNodeTypes } from '../ir'

import type {
  RootIRNode as VaporRootIRNode,
  CompilerOptions as VaporCompilerOptions
} from '@vue-vapor/compiler-vapor'
import type { CompilerOptions } from '../compile'
import type { RootNode, RootIRNode } from '../ir'

export const DEFAULT_OPTIONS: CompilerOptions = {
  prefixIdentifiers: true
}
export const DEFAULT_VAPOR_COMPILER_OPTIONS = DEFAULT_OPTIONS as VaporCompilerOptions

export function makeCompile(options: CompilerOptions = {}) {
  return (
    source: string,
    overrideOptions: CompilerOptions = {}
  ): {
    ast: RootNode
    ir: RootIRNode
    code: string
    helpers: Set<string>
    vaporHelpers: Set<string>
  } => {
    const svelteAst = parse(source)

    const ast: RootNode = {
      type: IRNodeTypes.ROOT,
      children: svelteAst.html.children || [],
      source,
      components: [],
      directives: [],
      helpers: new Set(),
      temps: 0
    }

    const ir = transform(ast, {
      ...DEFAULT_OPTIONS,
      ...options,
      ...overrideOptions
    })
    const { code, helpers, vaporHelpers } = generate(ir as unknown as VaporRootIRNode, {
      ...DEFAULT_OPTIONS,
      ...options,
      ...overrideOptions
    })

    return { ast, ir, code, helpers, vaporHelpers }
  }
}
