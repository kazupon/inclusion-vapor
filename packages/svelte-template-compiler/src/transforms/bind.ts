// SPDX-License-Identifier: MIT
// Modifier: kazuya kawaguchi (a.k.a. kazupon)
// Forked from `@vue/compiler-vapor`
// Author: Evan you (https://github.com/yyx990803) and Vapor team (https://github.com/orgs/vuejs/teams/vapor)
// Repository url: https://github.com/vuejs/core-vapor
// Code url: https://github.com/vuejs/core-vapor/blob/6608bb31973d35973428cae4fbd62026db068365/packages/compiler-vapor/src/transforms/vBind.ts

import {
  ErrorCodes,
  NodeTypes,
  createCompilerError,
  createSimpleExpression
} from '@vue-vapor/compiler-dom'
import { camelize, extend } from '@vue-vapor/shared'
import { isReservedProp, resolveExpression } from './utils.ts'

import type { SimpleExpressionNode } from '@vue-vapor/compiler-dom'
import type { TransformContext } from './context.ts'
import type { DirectiveTransform } from './types.ts'

// same-name shorthand - :arg is expanded to :arg="arg"
export function normalizeBindShorthand(
  arg: SimpleExpressionNode,
  context: TransformContext
): SimpleExpressionNode {
  if (arg.type !== NodeTypes.SIMPLE_EXPRESSION || !arg.isStatic) {
    // only simple expression is allowed for same-name shorthand
    context.options.onError(
      createCompilerError(ErrorCodes.X_V_BIND_INVALID_SAME_NAME_ARGUMENT, arg.loc)
    )
    return createSimpleExpression('', true, arg.loc)
  }

  const propName = camelize(arg.content)
  const exp = createSimpleExpression(propName, false, arg.loc)
  exp.ast = null // eslint-disable-line unicorn/no-null
  return exp
}

export const transformVBind: DirectiveTransform = (dir, _node, context) => {
  const { loc, modifiers } = dir
  let { exp } = dir
  let arg = dir.arg!
  const modifiersString = new Set(modifiers.map(s => s.content))

  if (!exp) {
    exp = normalizeBindShorthand(arg, context)
  }
  if (!exp.content.trim()) {
    context.options.onError(createCompilerError(ErrorCodes.X_V_BIND_NO_EXPRESSION, loc))
    exp = createSimpleExpression('', true, loc)
  }

  exp = resolveExpression(exp)
  arg = resolveExpression(arg)

  if (arg.isStatic && isReservedProp(arg.content)) {
    return
  }

  let camel = false
  if (modifiersString.has('camel')) {
    if (arg.isStatic) {
      arg = extend({}, arg, { content: camelize(arg.content) })
    } else {
      camel = true
    }
  }

  return {
    key: arg,
    value: exp,
    loc,
    runtimeCamelize: camel,
    modifier: modifiersString.has('prop') ? '.' : modifiersString.has('attr') ? '^' : undefined
  }
}
