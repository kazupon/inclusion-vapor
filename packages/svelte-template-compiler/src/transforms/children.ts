// SPDX-License-Identifier: MIT
// Modifier: kazuya kawaguchi (a.k.a. kazupon)
// Forked from `@vue/compiler-vapor`
// Author: Evan you (https://github.com/yyx990803) and Vapor team (https://github.com/orgs/vuejs/teams/vapor)
// Repository url: https://github.com/vuejs/core-vapor
// Code url: https://github.com/vuejs/core-vapor/blob/6608bb31973d35973428cae4fbd62026db068365/packages/compiler-vapor/src/transforms/transformChildren.ts

import { DynamicFlag, IRNodeTypes, isSvelteText } from '../ir/index.ts'
import { transformNode } from '../transform.ts'

import type { IRDynamicInfo, SvelteElement } from '../ir/index.ts'
import type { TransformContext } from './context.ts'
import type { NodeTransform } from './types.ts'

export const transformChildren: NodeTransform = (node, context) => {
  const isFragment =
    node.type === IRNodeTypes.ROOT ||
    node.type === 'Fragment' ||
    node.type === 'InlineComponent' ||
    node.type === 'Slot'
  if (node.type !== 'Element' && !isFragment) {
    return
  }
  processChildren(node, context, isFragment)
}

export function processChildren(
  node: Parameters<NodeTransform>[0],
  context: Parameters<NodeTransform>[1],
  isFragment: boolean
): ReturnType<NodeTransform> {
  const children = node.children || []

  // normalize svelte text
  for (const [index, child] of [...children].entries()) {
    if (isSvelteText(child) && !child.data.trim()) {
      child.data = ' '
      if (!index) {
        children.splice(0, 1)
      } else if (index === children.length) {
        children.splice(-1, 1)
      }
    }
  }

  for (const [i, child] of children.entries()) {
    const childContext = context.create(child, i)
    transformNode(childContext)

    if (isFragment) {
      childContext.reference()
      childContext.registerTemplate()

      if (
        !(childContext.dynamic.flags & DynamicFlag.NON_TEMPLATE) ||
        childContext.dynamic.flags & DynamicFlag.INSERT
      ) {
        context.block.returns.push(childContext.dynamic.id!)
      }
    } else {
      context.childrenTemplate.push(childContext.template)
    }

    context.dynamic.children[i] = childContext.dynamic
  }

  if (!isFragment) {
    processDynamicChildren(context as TransformContext<SvelteElement>)
  }
}

function processDynamicChildren(context: TransformContext<SvelteElement>) {
  let prevDynamics: IRDynamicInfo[] = []
  let hasStaticTemplate = false
  const children = context.dynamic.children

  for (const [index, child] of children.entries()) {
    if (child.flags & DynamicFlag.INSERT) {
      prevDynamics.push(child)
    }

    if (!(child.flags & DynamicFlag.NON_TEMPLATE)) {
      if (prevDynamics.length > 0) {
        if (hasStaticTemplate) {
          context.childrenTemplate[index - prevDynamics.length] = `<!>`

          prevDynamics[0].flags -= DynamicFlag.NON_TEMPLATE
          const anchor = (prevDynamics[0].anchor = context.increaseId())

          context.registerOperation({
            type: IRNodeTypes.INSERT_NODE,
            elements: prevDynamics.map(child => child.id!),
            parent: context.reference(),
            anchor
          })
        } else {
          context.registerOperation({
            type: IRNodeTypes.PREPEND_NODE,
            elements: prevDynamics.map(child => child.id!),
            parent: context.reference()
          })
        }
        prevDynamics = []
      }
      hasStaticTemplate = true
    }
  }

  if (prevDynamics.length > 0) {
    context.registerOperation({
      type: IRNodeTypes.INSERT_NODE,
      elements: prevDynamics.map(child => child.id!),
      parent: context.reference()
    })
  }
}
