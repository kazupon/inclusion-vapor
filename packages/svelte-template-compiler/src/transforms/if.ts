// SPDX-License-Identifier: MIT
// Modifier: kazuya kawaguchi (a.k.a. kazupon)
// Forked from `@vue/compiler-vapor`
// Author: Evan you (https://github.com/yyx990803) and Vapor team (https://github.com/orgs/vuejs/teams/vapor)
// Repository url: https://github.com/vuejs/core-vapor
// Code url: https://github.com/vuejs/core-vapor/blob/6608bb31973d35973428cae4fbd62026db068365/packages/compiler-vapor/src/transforms/vIf.ts

import { DynamicFlag, IRNodeTypes, isIfBlockOnElseBlock, isIfBlockOnTop } from '../ir/index.ts'
import { processChildren } from './children.ts'
import { newBlock, resolveSimpleExpression } from './utils.ts'

import type { BlockIRNode, IfIRNode, SvelteIfBlock, SvelteTemplateNode } from '../ir/index.ts'
import type { TransformContext } from './context.ts'
import type { NodeTransform } from './types.ts'

/**
 * transform vapor v-if for svelte {#if}
 * https://svelte.dev/docs/logic-blocks#if
 */
export const transformVIf: NodeTransform = (node, context) => {
  if (node.type === 'IfBlock' && !node.elseif) {
    return processIf(node as SvelteIfBlock, context as TransformContext<SvelteIfBlock>, [])
  }
}

export function processIf(
  node: SvelteIfBlock,
  context: TransformContext<SvelteIfBlock>,
  exitFns: (() => void)[],
  ifNode?: IfIRNode,
  vFor?: boolean
): (() => void)[] {
  let id = -1
  if (isIfBlockOnTop(node) && ifNode === undefined) {
    context.dynamic.flags |= DynamicFlag.NON_TEMPLATE | DynamicFlag.INSERT
    id = context.reference()
  }

  const condition = resolveSimpleExpression(node, context)
  if (vFor) {
    condition.content = `Array.from(${condition.content}).length`
  }
  const [positive, onExit] = createIfBranch(node, context)
  const operation: IfIRNode = {
    type: IRNodeTypes.IF,
    id,
    condition,
    positive,
    once: context.inVOnce
  }

  // set new BlockIRNode to negative
  if (isIfBlockOnElseBlock(node) && ifNode) {
    ifNode.negative = operation
  }

  // process children
  processChildren(node, context, true)

  exitFns.push(() => {
    onExit()
    if (isIfBlockOnTop(node)) {
      // if `#if` blocks is top, register operation
      context.registerOperation(operation)
    }
  })

  // process `:else if` or `:else` block
  if (node.else?.type === 'ElseBlock') {
    const ifBlock = node.else.children.find(child => child.type === 'IfBlock') as SvelteIfBlock
    if (ifBlock == undefined) {
      // for `:else` block
      const [negative, onExit] = createIfBranch(node.else, context)
      operation.negative = negative
      processChildren(node.else, context, true)
      exitFns.push(() => onExit())
    } else {
      // for `:else if` block
      return processIf(ifBlock, context, exitFns, operation)
    }
  }

  return exitFns
}

function createIfBranch(
  node: SvelteTemplateNode,
  context: TransformContext<SvelteTemplateNode>
): [BlockIRNode, () => void] {
  context.node = node // TODO: Should be wrapped?

  const branch: BlockIRNode = newBlock(node)
  const exitBlock = context.enterBlock(branch)
  context.reference()

  return [branch, exitBlock]
}
