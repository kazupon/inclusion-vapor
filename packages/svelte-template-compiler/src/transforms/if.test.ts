import { NodeTypes } from '@vue-vapor/compiler-dom'
import { compile as vaporCompile } from '@vue-vapor/compiler-vapor'
import { expect, test } from 'vitest'
import { IRNodeTypes } from '../ir/index.ts'
import { makeCompile } from './_utils.ts'
import { transformVBind } from './bind.ts'
import { transformChildren } from './children.ts'
import { transformComment } from './comment.ts'
import { transformElement } from './element.ts'
import { transformVIf } from './if.ts'
import { transformText } from './text.ts'

import type { IfIRNode } from '../ir/index.ts'

const compileWithVIf = makeCompile({
  prefixIdentifiers: false,
  nodeTransforms: [
    transformElement,
    transformChildren,
    transformText,
    transformComment,
    transformVIf
  ],
  directiveTransforms: {
    bind: transformVBind
  }
})

test('basic', () => {
  const source1 = `{#if ok}<div>{msg}</div>{/if}`
  const source2 = `<div v-if="ok">{{msg}}</div>`

  const { code, vaporHelpers, ir, helpers } = compileWithVIf(source1)
  const expectedResult = vaporCompile(source2)

  expect(code).toMatchSnapshot('received')
  expect(expectedResult.code).toMatchSnapshot('expected')

  expect(vaporHelpers).contains('createIf')
  expect(helpers.size).toBe(0)
  expect(ir.template).toEqual(['<div></div>'])
  expect(ir.block.operation).toMatchObject([
    {
      type: IRNodeTypes.IF,
      id: 0,
      condition: {
        type: NodeTypes.SIMPLE_EXPRESSION,
        content: 'ok',
        isStatic: false
      },
      positive: {
        type: IRNodeTypes.BLOCK,
        dynamic: {
          children: [{ template: 0 }]
        }
      }
    }
  ])
  expect(ir.block.returns).toEqual([0])

  expect(ir.block.dynamic).toMatchObject({
    children: [{ id: 0 }]
  })

  expect(ir.block.effect).toEqual([])
  expect((ir.block.operation[0] as IfIRNode).positive.effect).lengthOf(1)
})

test('#if + :else', () => {
  const source1 = `{#if ok}<div></div>{:else}<p></p>{/if}`
  const source2 = `<div v-if="ok"/><p v-else/>`

  const { code, ir, vaporHelpers, helpers } = compileWithVIf(source1)
  const expectedResult = vaporCompile(source2)

  expect(code).toMatchSnapshot('received')
  expect(expectedResult.code).toMatchSnapshot('expected')

  expect(ir.template).toEqual(['<div></div>', '<p></p>'])
  expect(vaporHelpers).contains('createIf')
  expect(ir.block.effect).lengthOf(0)
  expect(helpers).lengthOf(0)
  expect(ir.block.operation).toMatchObject([
    {
      type: IRNodeTypes.IF,
      id: 0,
      condition: {
        type: NodeTypes.SIMPLE_EXPRESSION,
        content: 'ok',
        isStatic: false
      },
      positive: {
        type: IRNodeTypes.BLOCK,
        dynamic: {
          children: [{ template: 0 }]
        }
      },
      negative: {
        type: IRNodeTypes.BLOCK,
        dynamic: {
          children: [{ template: 1 }]
        }
      }
    }
  ])
  expect(ir.block.returns).toEqual([0])
})

test('#if + :else-if', () => {
  const source1 = `{#if ok}<div></div>{:else if orNot}<p></p>{/if}`
  const source2 = `<div v-if="ok"/><p v-else-if="orNot"/>`

  const { code, ir } = compileWithVIf(source1)
  const expectedResult = vaporCompile(source2)

  expect(code).toMatchSnapshot('received')
  expect(expectedResult.code).toMatchSnapshot('expected')

  expect(ir.template).toEqual(['<div></div>', '<p></p>'])
  expect(ir.block.operation).toMatchObject([
    {
      type: IRNodeTypes.IF,
      id: 0,
      condition: {
        type: NodeTypes.SIMPLE_EXPRESSION,
        content: 'ok',
        isStatic: false
      },
      positive: {
        type: IRNodeTypes.BLOCK,
        dynamic: {
          children: [{ template: 0 }]
        }
      },
      negative: {
        type: IRNodeTypes.IF,
        condition: {
          type: NodeTypes.SIMPLE_EXPRESSION,
          content: 'orNot',
          isStatic: false
        },
        positive: {
          type: IRNodeTypes.BLOCK,
          dynamic: {
            children: [{ template: 1 }]
          }
        }
      }
    }
  ])
  expect(ir.block.returns).toEqual([0])
})

test('#if + :else-if + :else', () => {
  const source1 = `{#if ok}<div></div>{:else if orNot}<p></p>{:else}<span>fine</span>{/if}`
  const source2 = `<div v-if="ok"/><p v-else-if="orNot"/><span v-else>fine</span>`

  const { code, ir } = compileWithVIf(source1)
  const expectedResult = vaporCompile(source2)

  expect(code).toMatchSnapshot('received')
  expect(expectedResult.code).toMatchSnapshot('expected')

  expect(ir.template).toEqual(['<div></div>', '<p></p>', '<span>fine</span>'])
  expect(ir.block.returns).toEqual([0])
  expect(ir.block.operation).toMatchObject([
    {
      type: IRNodeTypes.IF,
      id: 0,
      positive: {
        type: IRNodeTypes.BLOCK,
        dynamic: {
          children: [{ template: 0 }]
        }
      },
      negative: {
        type: IRNodeTypes.IF,
        positive: {
          type: IRNodeTypes.BLOCK,
          dynamic: {
            children: [{ template: 1 }]
          }
        },
        negative: {
          type: IRNodeTypes.BLOCK,
          dynamic: {
            children: [{ template: 2 }]
          }
        }
      }
    }
  ])
})

test('dedupe same template', () => {
  const source1 = `{#if ok}<div>hello</div>{/if}{#if ok}<div>hello</div>{/if}`
  const source2 = `<div v-if="ok">hello</div><div v-if="ok">hello</div>`

  const { code, ir } = compileWithVIf(source1)
  const expectedResult = vaporCompile(source2)

  expect(code).toMatchSnapshot('received')
  expect(expectedResult.code).toMatchSnapshot('expected')

  expect(ir.template).toEqual(['<div>hello</div>'])
  expect(ir.block.returns).toEqual([0, 3])
})

test('comment between blocks', () => {
  const source1 = `{#if ok}
  <div></div>
  <!--foo-->
{:else if orNot}
  <p></p>
  <!--bar-->
{:else}
  fine
{/if}`
  const source2 = `<div v-if="ok"/>
<!--foo-->
<p v-else-if="orNot"/>
<!--bar-->
<template v-else>fine</template>
`
  const { code, ir } = compileWithVIf(source1)
  const expectedResult = vaporCompile(source2)

  expect(code).toMatchSnapshot('received')
  expect(expectedResult.code).toMatchSnapshot('expected')

  // TODO: should normalize whitespace or line breaks for Svelte Text node
  for (const template of ['<div></div>', '<!--foo-->', '<p></p>', '<!--bar-->', 'fine']) {
    expect(ir.template).toContain(template)
  }
  // expect(ir.template).toEqual(['<div></div>', '<!--foo-->', '<p></p>', '<!--bar-->', 'fine'])
})

test('nested #if', () => {
  const source1 = `{#if ok}
  {#if nested}
    <span>nested</span>
  {:else if nestedElse}
    <span>nestedElseIf</span>
  {:else}
    <span>nestedElse</span>
  {/if}
{:else}
  fine
{/if}`
  const source2 = `<template v-if="ok">
  <span v-if="nested">nested</span>
  <span v-else-if="nestedElse">nestedElseIf</span>
  <span v-else>nestedElse</span>
</template>
<template v-else>fine</template>`
  const { code, ir } = compileWithVIf(source1)
  const expectedResult = vaporCompile(source2)

  expect(code).toMatchSnapshot('received')
  expect(expectedResult.code).toMatchSnapshot('expected')

  expect(ir.template).toEqual([
    '<span>nested</span>',
    '<span>nestedElseIf</span>',
    '<span>nestedElse</span>',
    'fine'
  ])
  expect(ir.block.returns).toEqual([0])
  expect(ir.block.operation).toMatchObject([
    {
      type: IRNodeTypes.IF,
      id: 0,
      condition: {
        type: NodeTypes.SIMPLE_EXPRESSION,
        content: 'ok',
        isStatic: false
      },
      positive: {
        type: IRNodeTypes.BLOCK,
        operation: [
          {
            type: IRNodeTypes.IF,
            condition: {
              type: NodeTypes.SIMPLE_EXPRESSION,
              content: 'nested',
              isStatic: false
            },
            positive: {
              type: IRNodeTypes.BLOCK,
              dynamic: {
                children: [{ template: 0 }]
              }
            },
            negative: {
              type: IRNodeTypes.IF,
              condition: {
                type: NodeTypes.SIMPLE_EXPRESSION,
                content: 'nestedElse',
                isStatic: false
              },
              positive: {
                type: IRNodeTypes.BLOCK,
                dynamic: {
                  children: [{ template: 1 }]
                }
              },
              negative: {
                type: IRNodeTypes.BLOCK,
                dynamic: {
                  children: [{ template: 2 }]
                }
              }
            }
          }
        ]
      },
      negative: {
        type: IRNodeTypes.BLOCK,
        dynamic: {
          children: [{ template: 3 }]
        }
      }
    }
  ])
})

test.todo('component #if', () => {
  const source = `{#if ok}
  <MyComponent />
{/if}`
  expect(source).toBe('todo')
})

test('prefixIdentifiers: true', () => {
  const source1 = `{#if ok}<div>{msg}</div>{/if}`
  const source2 = `<div v-if="ok">{{msg}}</div>`

  const { code, vaporHelpers, ir, helpers } = compileWithVIf(source1, { prefixIdentifiers: true })
  const expectedResult = vaporCompile(source2)

  expect(code).toMatchSnapshot('received')
  expect(expectedResult.code).toMatchSnapshot('expected')

  expect(vaporHelpers).contains('createIf')
  expect(helpers.size).toBe(0)
  expect(ir.template).toEqual(['<div></div>'])
  expect(ir.block.operation).toMatchObject([
    {
      type: IRNodeTypes.IF,
      id: 0,
      condition: {
        type: NodeTypes.SIMPLE_EXPRESSION,
        content: 'ok',
        isStatic: false
      },
      positive: {
        type: IRNodeTypes.BLOCK,
        dynamic: {
          children: [{ template: 0 }]
        }
      }
    }
  ])
  expect(ir.block.returns).toEqual([0])

  expect(ir.block.dynamic).toMatchObject({
    children: [{ id: 0 }]
  })

  expect(ir.block.effect).toEqual([])
  expect((ir.block.operation[0] as IfIRNode).positive.effect).lengthOf(1)
})
