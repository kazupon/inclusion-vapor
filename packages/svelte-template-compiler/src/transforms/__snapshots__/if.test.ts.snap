// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`#if + :else > svelte 1`] = `
"import { createIf as _createIf, template as _template } from 'vue/vapor';
const t0 = _template("<div></div>")
const t1 = _template("<p></p>")

export function render(_ctx) {
  const n0 = _createIf(() => (ok), () => {
    const n2 = t0()
    return n2
  }, () => {
    const n4 = t1()
    return n4
  })
  return n0
}"
`;

exports[`#if + :else > vue 1`] = `
"import { createIf as _createIf, template as _template } from 'vue/vapor';
const t0 = _template("<div></div>")
const t1 = _template("<p></p>")

export function render(_ctx) {
  const n0 = _createIf(() => (ok), () => {
    const n2 = t0()
    return n2
  }, () => {
    const n4 = t1()
    return n4
  })
  return n0
}"
`;

exports[`#if + :else-if + :else > svelte 1`] = `
"import { createIf as _createIf, template as _template } from 'vue/vapor';
const t0 = _template("<div></div>")
const t1 = _template("<p></p>")
const t2 = _template("<span>fine</span>")

export function render(_ctx) {
  const n0 = _createIf(() => (ok), () => {
    const n2 = t0()
    return n2
  }, () => _createIf(() => (orNot), () => {
    const n4 = t1()
    return n4
  }, () => {
    const n6 = t2()
    return n6
  }))
  return n0
}"
`;

exports[`#if + :else-if + :else > vue 1`] = `
"import { createIf as _createIf, template as _template } from 'vue/vapor';
const t0 = _template("<div></div>")
const t1 = _template("<p></p>")
const t2 = _template("<span>fine</span>")

export function render(_ctx) {
  const n0 = _createIf(() => (ok), () => {
    const n2 = t0()
    return n2
  }, () => _createIf(() => (orNot), () => {
    const n4 = t1()
    return n4
  }, () => {
    const n7 = t2()
    return n7
  }))
  return n0
}"
`;

exports[`#if + :else-if > svelte 1`] = `
"import { createIf as _createIf, template as _template } from 'vue/vapor';
const t0 = _template("<div></div>")
const t1 = _template("<p></p>")

export function render(_ctx) {
  const n0 = _createIf(() => (ok), () => {
    const n2 = t0()
    return n2
  }, () => _createIf(() => (orNot), () => {
    const n4 = t1()
    return n4
  }))
  return n0
}"
`;

exports[`#if + :else-if > vue 1`] = `
"import { createIf as _createIf, template as _template } from 'vue/vapor';
const t0 = _template("<div></div>")
const t1 = _template("<p></p>")

export function render(_ctx) {
  const n0 = _createIf(() => (ok), () => {
    const n2 = t0()
    return n2
  }, () => _createIf(() => (orNot), () => {
    const n4 = t1()
    return n4
  }))
  return n0
}"
`;

exports[`basic > svelte 1`] = `
"import { createTextNode as _createTextNode, insert as _insert, renderEffect as _renderEffect, setText as _setText, createIf as _createIf, template as _template } from 'vue/vapor';
const t0 = _template("<div></div>")

export function render(_ctx) {
  const n0 = _createIf(() => (ok), () => {
    const n3 = t0()
    const n2 = _createTextNode(() => [msg])
    _insert(n2, n3)
    _renderEffect(() => _setText(n3, msg))
    return n3
  })
  return n0
}"
`;

exports[`basic > vue 1`] = `
"import { renderEffect as _renderEffect, setText as _setText, createIf as _createIf, template as _template } from 'vue/vapor';
const t0 = _template("<div></div>")

export function render(_ctx) {
  const n0 = _createIf(() => (ok), () => {
    const n2 = t0()
    _renderEffect(() => _setText(n2, msg))
    return n2
  })
  return n0
}"
`;

exports[`comment between blocks > svelte 1`] = `
"import { createIf as _createIf, template as _template } from 'vue/vapor';
const t0 = _template("<div></div>")
const t1 = _template(" ")
const t2 = _template("<!--foo-->")
const t3 = _template("<p></p>")
const t4 = _template("<!--bar-->")
const t5 = _template("fine")

export function render(_ctx) {
  const n0 = _createIf(() => (ok), () => {
    const n2 = t0()
    const n3 = t1()
    const n4 = t2()
    return [n2, n3, n4]
  }, () => _createIf(() => (orNot), () => {
    const n6 = t3()
    const n7 = t1()
    const n8 = t4()
    return [n6, n7, n8]
  }, () => {
    const n10 = t5()
    return n10
  }))
  return n0
}"
`;

exports[`comment between blocks > vue 1`] = `
"import { createIf as _createIf, template as _template } from 'vue/vapor';
const t0 = _template("<div></div>")
const t1 = _template("<!--foo-->")
const t2 = _template("<p></p>")
const t3 = _template("<!--bar-->")
const t4 = _template("fine")

export function render(_ctx) {
  const n0 = _createIf(() => (ok), () => {
    const n2 = t0()
    return n2
  }, () => _createIf(() => (orNot), () => {
    const n5 = t1()
    const n6 = t2()
    return [n5, n6]
  }, () => {
    const n10 = t3()
    const n11 = t4()
    return [n10, n11]
  }))
  return n0
}"
`;

exports[`dedupe same template > svelte 1`] = `
"import { createIf as _createIf, template as _template } from 'vue/vapor';
const t0 = _template("<div>hello</div>")

export function render(_ctx) {
  const n0 = _createIf(() => (ok), () => {
    const n2 = t0()
    return n2
  })
  const n3 = _createIf(() => (ok), () => {
    const n5 = t0()
    return n5
  })
  return [n0, n3]
}"
`;

exports[`dedupe same template > vue 1`] = `
"import { createIf as _createIf, template as _template } from 'vue/vapor';
const t0 = _template("<div>hello</div>")

export function render(_ctx) {
  const n0 = _createIf(() => (ok), () => {
    const n2 = t0()
    return n2
  })
  const n3 = _createIf(() => (ok), () => {
    const n5 = t0()
    return n5
  })
  return [n0, n3]
}"
`;

exports[`nested #if > svelte 1`] = `
"import { createIf as _createIf, template as _template } from 'vue/vapor';
const t0 = _template("<span>nested</span>")
const t1 = _template("<span>nestedElseIf</span>")
const t2 = _template("<span>nestedElse</span>")
const t3 = _template("fine")

export function render(_ctx) {
  const n0 = _createIf(() => (ok), () => {
    const n2 = _createIf(() => (nested), () => {
      const n4 = t0()
      return n4
    }, () => _createIf(() => (nestedElse), () => {
      const n6 = t1()
      return n6
    }, () => {
      const n8 = t2()
      return n8
    }))
    return n2
  }, () => {
    const n10 = t3()
    return n10
  })
  return n0
}"
`;

exports[`nested #if > vue 1`] = `
"import { createIf as _createIf, template as _template } from 'vue/vapor';
const t0 = _template("<span>nested</span>")
const t1 = _template("<span>nestedElseIf</span>")
const t2 = _template("<span>nestedElse</span>")
const t3 = _template("fine")

export function render(_ctx) {
  const n0 = _createIf(() => (ok), () => {
    const n2 = _createIf(() => (nested), () => {
      const n4 = t0()
      return n4
    }, () => _createIf(() => (nestedElse), () => {
      const n6 = t1()
      return n6
    }, () => {
      const n9 = t2()
      return n9
    }))
    return n2
  }, () => {
    const n12 = t3()
    return n12
  })
  return n0
}"
`;

exports[`prefixIdentifiers: true > svelte 1`] = `
"import { createTextNode as _createTextNode, insert as _insert, renderEffect as _renderEffect, setText as _setText, createIf as _createIf, template as _template } from 'vue/vapor';
const t0 = _template("<div></div>")

export function render(_ctx) {
  const n0 = _createIf(() => (_ctx.ok), () => {
    const n3 = t0()
    const n2 = _createTextNode(() => [_ctx.msg])
    _insert(n2, n3)
    _renderEffect(() => _setText(n3, _ctx.msg))
    return n3
  })
  return n0
}"
`;

exports[`prefixIdentifiers: true > vue 1`] = `
"import { renderEffect as _renderEffect, setText as _setText, createIf as _createIf, template as _template } from 'vue/vapor';
const t0 = _template("<div></div>")

export function render(_ctx) {
  const n0 = _createIf(() => (ok), () => {
    const n2 = t0()
    _renderEffect(() => _setText(n2, msg))
    return n2
  })
  return n0
}"
`;
